rasa/core/policies/ted_policy.py

```diff
@@ -579,10 +583,58 @@ def predict_action_probabilities(
         if self.config[LOSS_TYPE] == SOFTMAX and self.config[RANKING_LENGTH] > 0:
             confidence = train_utils.normalize(confidence, self.config[RANKING_LENGTH])
 
+        optional_events = self._create_optional_event_for_entities(
+            output, interpreter, tracker
+        )
+
         return self._prediction(
-            confidence.tolist(), is_end_to_end_prediction=is_e2e_prediction
+            confidence.tolist(),
+            is_end_to_end_prediction=is_e2e_prediction,
+            optional_events=optional_events,
+        )
+
+    def _create_optional_event_for_entities(
+        self,
+        prediction_output: Dict[Text, tf.Tensor],
+        interpreter: NaturalLanguageInterpreter,
+        tracker: DialogueStateTracker,
+    ) -> Optional[List[Event]]:
+        if not self.config[ENTITY_RECOGNITION]:
+            # entity recognition is not turned on, no entities can be predicted
+            return None
+
+        (
+            predicted_tags,
+            confidence_values,
+        ) = rasa.utils.train_utils.entity_label_to_tags(
+            prediction_output, self._entity_tag_specs
         )
 
+        if ENTITY_ATTRIBUTE_TYPE not in predicted_tags:
+            # no entities detected
+            return None
+
+        # find last user uttered event as the predicted entities belong to
+        # that utterance
+        if tracker.latest_action_name == ACTION_LISTEN_NAME:
+            last_user_utterance = tracker.latest_message
+        else:
+            return None
```

Comment:
I think
```
if tracker.latest_action_name == ACTION_LISTEN_NAME:
    return
```
should be the first check to save calculation and extraction of entity labels etc
actually, in this case TED itself should return smth like `None`?

Reply:
I guess you mean `!=` instead of `==`.

Reply:
yes